exporter/elasticsearchexporter/data-stream-router.go

```diff
@@ -0,0 +1,65 @@
+)
+
+// routeLogRecord returns the name of the index to send the log record to according to data stream routing attributes.
+// It searches for the routing attributes on the log record, scope, and resource.
+// It creates missing routing attributes on the log record if they are not found.
+func routeLogRecord(
+	record *plog.LogRecord,
+	scope pcommon.InstrumentationScope,
+	resource pcommon.Resource,
+) string {
+	dataSet := ensureAttribute(dataStreamDataset, defaultDataStreamDataset, record.Attributes(), scope.Attributes(), resource.Attributes())
+	namespace := ensureAttribute(dataStreamNamespace, defaultDataStreamNamespace, record.Attributes(), scope.Attributes(), resource.Attributes())
+	record.Attributes().PutStr(dataStreamType, defaultDataStreamTypeLogs)
+	return fmt.Sprintf("%s-%s-%s", defaultDataStreamTypeLogs, dataSet, namespace)
+}
+
+// routeDataPoint returns the name of the index to send the data point to according to data stream routing attributes.
+// It searches for the routing attributes on the data point, scope, and resource.
+// It creates missing routing attributes on the data point if they are not found.
+func routeDataPoint(
+	dataPoint pmetric.NumberDataPoint,
+	scope pcommon.InstrumentationScope,
+	resource pcommon.Resource,
+) string {
+	dataSet := ensureAttribute(dataStreamDataset, defaultDataStreamDataset, dataPoint.Attributes(), scope.Attributes(), resource.Attributes())
+	namespace := ensureAttribute(dataStreamNamespace, defaultDataStreamNamespace, dataPoint.Attributes(), scope.Attributes(), resource.Attributes())
+	dataPoint.Attributes().PutStr(dataStreamType, defaultDataStreamTypeMetrics)
+	return fmt.Sprintf("%s-%s-%s", defaultDataStreamTypeMetrics, dataSet, namespace)
+}
+
+// routeSpan returns the name of the index to send the span to according to data stream routing attributes.
+// It searches for the routing attributes on the span, scope, and resource.
+// It creates missing routing attributes on the span if they are not found.
+func routeSpan(
+	span ptrace.Span,
+	scope pcommon.InstrumentationScope,
+	resource pcommon.Resource,
+) string {
+	dataSet := ensureAttribute(dataStreamDataset, defaultDataStreamDataset, span.Attributes(), scope.Attributes(), resource.Attributes())
+	namespace := ensureAttribute(dataStreamNamespace, defaultDataStreamNamespace, span.Attributes(), scope.Attributes(), resource.Attributes())
+	span.Attributes().PutStr(dataStreamType, defaultDataStreamTypeTraces)
+	return fmt.Sprintf("%s-%s-%s", defaultDataStreamTypeTraces, dataSet, namespace)
+}
+
+func ensureAttribute(attributeName string, defaultValue string, recordAttributes, scopeAttributes, resourceAttributes pcommon.Map) string {
+	// Fetch value according to precedence and default.
+	value := getFromAttributesNew(attributeName, defaultValue, recordAttributes, scopeAttributes, resourceAttributes)
+
+	// Always set the value on the record, as record attributes have the highest precedence.
```

Comment:
```suggestion
	// Set the attribute in case the default value is used.
	//
	// Always set the value on the record, as record attributes have the highest precedence.
```

I think? Should we only do this `if value == defaultValue`?

Reply:
Scope attributes are *not* read in the encodeModel, meaning that only Resource and DataPoint attributes are added to the Document. This is aligned with the behavior of apm-data, where scope attributes are only read for data_stream.* and anything else is ignored, see [code](https://github.com/elastic/apm-data/blob/0e072ceaadae2ef8f61d75c3a27c47cc5ce81257/input/otlp/metadata.go#L437). Let me know if you find this behavior problematic.

> Should we only do this if value == defaultValue

Therefore, the problem of doing this is that if data_stream.* is present in scope attributes, they will be missing from the resulting doc.

Reply:
I see, thanks. Seems odd, but not something we need to change right now.